src/main/java/net/neoforged/neoforge/event/BlockEntityTypeAddBlocksEvent.java

```diff
@@ -0,0 +1,121 @@
+
+    private BlockEntityTypeAddBlocksEvent(ResourceKey<BlockEntityType<?>> blockEntityTypeResourceKey, BlockEntityType<?> blockEntityType) {
+        this.blockEntityTypeResourceKey = blockEntityTypeResourceKey;
+        this.blockEntityType = blockEntityType;
+        this.currentValidBlocks = new HashSet<>(blockEntityType.getValidBlocks());
+    }
+
+    /**
+     * The {@link ResourceKey} that this current {@link BlockEntityType} was registered under.
+     */
+    public ResourceKey<BlockEntityType<?>> getBlockEntityTypeResourceKey() {
+        return blockEntityTypeResourceKey;
+    }
+
+    /**
+     * The current {@link BlockEntityType}.
+     */
+    public BlockEntityType<?> getBlockEntityType() {
+        return blockEntityType;
+    }
+
+    /**
+     * An immutable list of the list of valid blocks that will be assigned to replace {@link BlockEntityType#validBlocks}'s value.
+     */
+    public Set<Block> getCurrentValidBlocks() {
+        return Collections.unmodifiableSet(this.currentValidBlocks);
+    }
+
+    /**
+     * Will add the given block to the list of valid blocks we will assign to the current {@link BlockEntityType}.
+     * Make sure the given block is the same or derived from the highest common class of all the entries in valid blocks.
+     * <p>
+     * Example: If the valid blocks list has {@link StandingSignBlock} entry and {@link WallSignBlock} entry, the common class is SignBlock
+     * and the given block must be a {@link SignBlock} or has {@link SignBlock} as a child class in its hierarchy.
+     * 
+     * @param block The block to add as a valid block for the current {@link BlockEntityType}
+     */
+    public void addValidBlock(Block block) {
+        if (this.baseClass.get() == null || this.baseClass.get().isAssignableFrom(block.getClass())) {
+            this.currentValidBlocks.add(block);
+        } else {
+            throw new IllegalArgumentException("Given block " + block + " does not derive from existing valid block's common superclass of " + this.baseClass);
+        }
+    }
+
+    @Nullable
+    private Class<?> getCommonSuperClassForExistingValidBlocks() {
+        Class<?> calculatedBaseClass = null;
+
+        for (Block existingBlock : this.currentValidBlocks) {
+            if (calculatedBaseClass != null) {
+                calculatedBaseClass = findClosestCommonSuper(calculatedBaseClass, existingBlock.getClass());
+            } else {
+                calculatedBaseClass = existingBlock.getClass();
+            }
+        }
+
+        return calculatedBaseClass;
+    }
+
+    private static Class<?> findClosestCommonSuper(Class<?> superClass, Class<?> childClass) {
+        while (!superClass.isAssignableFrom(childClass)) {
+            superClass = superClass.getSuperclass();
+        }
+        return superClass;
+    }
+
+    public static void fireBlockEntityTypeValidAdditions() {
```

Comment:
Why is this in the event class and not in the place the event is fired?

Reply:
Not sure that is a good idea as I don't think it is clean to stuff this entire method into GameData. In fact, we already call external methods in gameData for cleanness.
![image](https://github.com/neoforged/NeoForge/assets/40846040/a189cd05-c5cf-4c9a-9671-91aefa37b352)

Reply:
At least make this method internal then.